File: Top200/108-SortedArrayToBST.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	return constructRecur(nums, 0, len(nums)-1)
}

func constructRecur(nums []int, left, right int) *TreeNode {
	if left > right {
		return nil
	}

	mid := left + ((right - left) >> 1)
	root := &TreeNode{
		Val: nums[mid],
	}

	root.Left = constructRecur(nums, left, mid-1)
	root.Right = constructRecur(nums, mid+1, right)

	return root
}

File: Top200/108-SortedArrayToBST.java
package Top200;

class SortedArrayToBST {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }

    public TreeNode sortedArrayToBST(int[] nums) {
        return recur(nums, 0, nums.length - 1);
    }

    public TreeNode recur(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }

        int mid = left + ((right - left) >> 1);
        TreeNode root = new TreeNode(nums[mid]);
        root.left = recur(nums, left, mid - 1);
        root.right = recur(nums, mid + 1, right);

        return root;
    }
    
}
